kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+    class CustomHelpFormatter(argparse.RawTextHelpFormatter):
+        pass
+
+    parser = argparse.ArgumentParser(
+        description='KCIDB Match Tool',
+        formatter_class=CustomHelpFormatter,
+        epilog='''\
+Usage examples:
+
+export -x DB_OPTS="postgresql:host=127.0.0.1 port=5432 sslmode=disable
+dbname=playground_kcidb user=helen.koike@collabora.com"
+
+# Update patterns
+kcidb-query -i "kernelci_api:70d17807303641a9d6d2a8aeb1aee829221cefcf"
+-d "$DB_OPTS" | ./kcidb-match.py --update-patterns
+
+# Generate incidents
+kcidb-query -t "maestro:6690dbfc7488a1b744200e82" -d "$DB_OPTS"
+--parents | ./kcidb-match.py --generate-incidents
+
+# Check test ID
+cat issue.json | ./kcidb-match.py --check_test_id
+"maestro:6690dbfc7488a1b744200e82" -d "$DB_OPTS"
+
+# Check build ID
+cat issue.json | ./kcidb-match.py --check_build_id
+"maestro:6690dbfc7488a1b744200e82" -d "$DB_OPTS"
+'''
+    )
+
+    parser.add_argument('--update-patterns', action='store_true',
+                        help='Update patterns from issues. Other '
+                             'arguments are ignored when used. Expects '
+                             'KCIDB-IO object with issues via stdin.')
+
+    parser.add_argument('--generate-incidents', action='store_true',
+                        help='Generate incidents for matched issues. '
+                             'Expects KCIDB-IO object with build and/or '
+                             'test via stdin.')
+
+    parser.add_argument('--ignore-db', action='store_true',
+                        help='Ignore the database and generate incidents '
+                             'based on the issues field in the KCIDB-IO '
+                             'object via stdin.')
+
+    parser.add_argument('--check_test_id', type=str,
+                        help='Test ID to check. Requires --db_conn. '
+                             'Implies --ignore-db. Expects KCIDB-IO '
+                             'object with issues via stdin.')
+
+    parser.add_argument('--check_build_id', type=str,
+                        help='Build ID to check. Requires --db_conn. '
+                             'Implies --ignore-db. '
+                             'Expects KCIDB-IO object with issues via stdin.')
+
+    parser.add_argument('-d', '--db_conn', type=str,
+                        help='Database connection string for kcidb-query.'
+                             'Required with --check_test_id or '
+                             '--check_build_id.')
+
+    args = parser.parse_args()
+
+    if args.check_test_id and args.check_build_id:
+        parser.error("Cannot use both --check_test_id and --check_build_id")
+
+    if (args.check_test_id or args.check_build_id) and not args.db_conn:
+        parser.error("--db_conn is required when using --check_test_id or "
+                     "--check_build_id")
+
+    if args.check_test_id or args.check_build_id:
+        args.ignore_db = True
+
+    return args
+
+
+def main():
```

Comment:
we need to re-think about this command line tool, maybe it could be simpler as:

cat issue.json | kcidb-match # nothing to output
cat tests.json | kcidb-match # and this would output the incidents that was generated

we could also have

kcidb-match --list-patterns # that would list all the saved objects inside patterns.db
kcidb-match --clear-db # that would erase pattern.db

but those two are not a priority, we can do this by navigating patterns.db with other tools

Reply:
A thought:
@helen-fornazier Do you think this tool is going to be used as a command line tool?
The current use of this tool is with the subscription module.
If we are going to expand the use cases of this tool, then only we should spend time refactoring the whole `main` function.

Reply:
For command line tool we would only use it as developers for debug, a normal user wouldn't use it I would say

Reply:
Thanks for clarifying.
As this tool is a high priority and we want to merge this asap, I won't spend time refactoring the main function atm. We can follow this up later if required.